# Applied Data Science Capstone - Notebook Week 3


__Segmenting and Clustering Neighborhoods in Toronto__


### Section #1 - Prepare neighborhood data

<span style="color:darkred">*Import the dependencies*</span>

In [1]:
import numpy as np
import pandas as pd

<span style="color:darkred">*Retrieve neighborhood data from Wikipedia*</span><br>
<span style="color:darkred">*We use a specific version of the page since it has been modified several times*</span>

In [2]:
df = pd.read_html('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&direction=prev&oldid=926287641')[0]
df

Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
283      M8Z         Etobicoke              Mimico NW
284      M8Z         Etobicoke     The Queensway West
285      M8Z         Etobicoke  Royal York South West
286      M8Z         Etobicoke         South of Bloor
287      M9Z      Not assigned           Not assigned

[288 rows x 3 columns]

<span style="color:darkred">*Fix the colums names*</span>

In [3]:
df.columns = ['PostalCode', 'Borough', 'Neighborhood']

<span style="color:darkred">*Filter out the post codes with no assigned borough*</span>

In [4]:
df = df[df['Borough'] != 'Not assigned'].reset_index(drop=True)
df

PostalCode           Borough              Neighborhood
0          M3A        North York                 Parkwoods
1          M4A        North York          Victoria Village
2          M5A  Downtown Toronto              Harbourfront
3          M5A  Downtown Toronto               Regent Park
4          M6A        North York          Lawrence Heights
..         ...               ...                       ...
206        M8Z         Etobicoke  Kingsway Park South West
207        M8Z         Etobicoke                 Mimico NW
208        M8Z         Etobicoke        The Queensway West
209        M8Z         Etobicoke     Royal York South West
210        M8Z         Etobicoke            South of Bloor

[211 rows x 3 columns]

<span style="color:darkred">*Assign unassigned neighborhoods to same as borough, if any*</span>

In [5]:
print('Before: {} neighborhoods not assigned'.format(len(df[df['Neighborhood'] == 'Not assigned'])))
df['Neighborhood'] = df['Neighborhood'].replace('Not assigned', df['Borough'])
print('After: {} neighborhoods not assigned'.format(len(df[df['Neighborhood'] == 'Not assigned'])))

Before: 1 neighborhoods not assigned
After: 0 neighborhoods not assigned


<span style="color:darkred">*Group together neighborhoods with same post code*</span>

In [6]:
df = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
df

PostalCode      Borough                                       Neighborhood
0          M1B  Scarborough                                     Rouge, Malvern
1          M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2          M1E  Scarborough                  Guildwood, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                          Cedarbrae
..         ...          ...                                                ...
98         M9N         York                                             Weston
99         M9P    Etobicoke                                          Westmount
100        M9R    Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...
101        M9V    Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...
102        M9W    Etobicoke                                          Northwest

[103 rows x 3 columns]

### Section #2 - Add geospacial data

<span style="color:darkred">*Define the function that retrieves the geospacial coordinates for a given postal code.*</span><br>
<span style="color:darkred">*We use geocoder.arcgis instead of geocoder.google as the latter doesn't seem to work. As a consequence, we will retrieve lattude and longitude coordinates that are slightly different from what is provided in the csv file but that shouldn't change the final clustering results.*</span>

In [7]:
import geocoder # import geocoder

def retrieve_geo_data(postal_code):
    # initialize your variable to None
    lat_lng_coords = None

    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng

    # found coordinates, return
    return lat_lng_coords[0], lat_lng_coords[1]

<span style="color:darkred">*Iterate through postal codes and add the corresponding latitudes and longitudes in the dataframe*</span>

In [8]:
# initialize
latitude = []
longitude = []

# iterate
for postalCode in df['PostalCode']:
    lat, long = retrieve_geo_data(postalCode)
    latitude.append(lat)
    longitude.append(long)
df['Latitude'] = latitude
df['Longitude'] = longitude

# show results
df

PostalCode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood  Latitude  Longitude  
0                                       Rouge, Malvern  43.81139  -79.19662  
1               Highland Creek, Rouge Hill, Port Union  43.78574  -79.15875  
2                    Guildwood, Morningside, West Hill  43.76575  -79.17470  
3                                               Woburn  43.76812  -79.21761  
4                                            Cedarbrae  43.76944  -79.23892  
..                                                 ...       ...        ...  
98                                              Weston  43.70507  -79.51804  
99                                           Westmount  43.69630  -79.52926  
100  Kingsview Village, Martin Grove Gardens, Richv...  43.68681  -79.55728  
101  Albion Gardens, Beaumond Heights, Humbergate, ...  43.74453  -79.58624  
102                                          Northwest  43.71174  -79.57941  

[103 rows x 5 columns]

### Section #3 - Clustering

##### <span style="color:darkred">1- Display Toronto neighborhoods</span>

<span style="color:darkred">*Determine Toronto coordinates that will be used as the center of the Folium map*</span>

In [9]:
from geopy.geocoders import Nominatim

geolocator = Nominatim(user_agent="capstone_explorer")
loc = geolocator.geocode("Toronto, ON")
toronto_coord = [loc.latitude, loc.longitude]

<span style="color:darkred">*Define a function that adds circle markers of specific color to the map*</spn>

In [10]:
import folium

def add_markers_to_map(df, color, add_label, map):
    for lat, lng, lbl in zip(df['Latitude'], df['Longitude'], df['Neighborhood']):
        label = folium.Popup("({}) {}".format(add_label, lbl), parse_html=True)
        folium.CircleMarker(
            [lat,lng], radius=5, popup=label,
            color=color, fill=True, fill_color=color, fill_opacity = 0.7,
            parse_html=False).add_to(map)

<span style="color:darkred">*Test the function: show the neiborhoods with a color corresponding to the borough*</span>

In [11]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create and center map
map_toronto = folium.Map(location=[loc.latitude, loc.longitude], zoom_start=11)

# define color range
borough_list = df['Borough'].unique().tolist()
colors_array = cm.rainbow(np.linspace(0, 1, len(borough_list)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# iterate on boroughs
for i, borough in enumerate(borough_list):
    add_markers_to_map(df[df['Borough']==borough].reset_index(),
        rainbow[i], "Borough='{}'".format(borough_list[i]), map_toronto)
    
# show map
map_toronto

##### <span style="color:darkred">2- Explore neighborhoods' nearby venues</span>

<span style="color:darkred">*We will cluster Toronto's neighborhoods according to the similarities of their nearby venues. We'll use the __proportion of venues of each type__ (category) found within __500 meters__ of the center of the neighborhoods as criterien for the K-means algorithm, and we will use __4 clusters__.*</span>

In [12]:
RADIUS = 500
NB_CLUSTERS = 4

<span style="color:darkred">*First we retrieve all venues from Foursquare*</span>

In [13]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

CLIENT_ID = 'X44GZE3TFHMZN4XSSRDPNQELI4O3WQXC2XBBI5VOWY52J3AE' # Foursquare ID
CLIENT_SECRET = 'BQSEQ22WUR5Z1P25YZN0GPBWK0T0PTKJ3C5HBW0YLWXP2A1C' # Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # default Foursquare API limit value

# iterate over neighboorhoods
venues_list=[]
for name, lat, lng in zip(df['Neighborhood'], df['Latitude'], df['Longitude']):
    # create the API request URL
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID, CLIENT_SECRET, VERSION, lat, lng, RADIUS, LIMIT) 
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    # return only relevant information for each nearby venue
    venues_list.append([(name, lat, lng, 
        v['venue']['name'], 
        v['venue']['location']['lat'], 
        v['venue']['location']['lng'],  
        v['venue']['categories'][0]['name']) for v in results])

# build dataframe
nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
nearby_venues.columns = ['Neighborhood Name', 'Neighborhood Latitude', 'Neighborhood Longitude', 
    'Venue',  'Venue Latitude', 'Venue Longitude', 'Venue Category']

# display results and size
nearby_venues

Neighborhood Name  Neighborhood Latitude  \
0                             Rouge, Malvern               43.81139   
1     Highland Creek, Rouge Hill, Port Union               43.78574   
2     Highland Creek, Rouge Hill, Port Union               43.78574   
3          Guildwood, Morningside, West Hill               43.76575   
4          Guildwood, Morningside, West Hill               43.76575   
...                                      ...                    ...   
2374                               Northwest               43.71174   
2375                               Northwest               43.71174   
2376                               Northwest               43.71174   
2377                               Northwest               43.71174   
2378                               Northwest               43.71174   

      Neighborhood Longitude                                    Venue  \
0                  -79.19662                                  Wendy’s   
1                  -79.15875                   Chris Effects Painting   
2                  -79.15875                    Royal Canadian Legion   
3                  -79.17470                 Homestead Roofing Repair   
4                  -79.17470              Heron Park Community Centre   
...                      ...                                      ...   
2374               -79.57941                             Petro-Canada   
2375               -79.57941                          Rexdale Hyundai   
2376               -79.57941  U-Haul Moving & Storage at Rexdale Blvd   
2377               -79.57941                            Fitness Two 4   
2378               -79.57941        Ali Baba's Middle Eastern Cuisine   

      Venue Latitude  Venue Longitude              Venue Category  
0          43.807448       -79.199056        Fast Food Restaurant  
1          43.784343       -79.163742  Construction & Landscaping  
2          43.782533       -79.163085                         Bar  
3          43.765140       -79.178663  Construction & Landscaping  
4          43.768867       -79.176958        Gym / Fitness Center  
...              ...              ...                         ...  
2374       43.714398       -79.581868                 Gas Station  
2375       43.713733       -79.578123             Auto Dealership  
2376       43.713500       -79.577307            Storage Facility  
2377       43.714482       -79.582903                         Gym  
2378       43.711704       -79.573288   Middle Eastern Restaurant  

[2379 rows x 7 columns]

In [14]:
print('There are {} uniques categories.'.format(len(nearby_venues['Venue Category'].unique())))

There are 266 uniques categories.


In [15]:
len(nearby_venues['Neighborhood Name'].unique())

101

<span style="color:darkred">*We can notice that Foursquare found venues around 101 neighborhoods out of the 103 initially listed, we will cluster the missing 2 neighborhoods together at the end.*</span>

<span style="color:darkred">*Now we create a dataframe with all the categories using __one-hot encoding__ then calculate the average of venues of each type per neighborhood using __groupby / mean__ aggregation method.*</span>

In [16]:
# create categories dataframe with one-hot encoding
venues_onehot = pd.get_dummies(nearby_venues[['Venue Category']], prefix="", prefix_sep="")
# add Neighborhood Name as first column
venues_onehot['Neighborhood Name'] = nearby_venues['Neighborhood Name']
new_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])
venues_onehot = venues_onehot[new_columns]
# calculate the number of each type of venue per neighborhood
venues_avg = venues_onehot.groupby('Neighborhood Name').mean().reset_index()
# show results
venues_avg

Neighborhood Name  ATM  \
0                             Adelaide, King, Richmond  0.0   
1                                            Agincourt  0.0   
2    Agincourt North, L'Amoreaux East, Milliken, St...  0.0   
3    Albion Gardens, Beaumond Heights, Humbergate, ...  0.0   
4                               Alderwood, Long Branch  0.0   
..                                                 ...  ...   
96                                     Willowdale West  0.0   
97                                              Woburn  0.0   
98                     Woodbine Gardens, Parkview Hill  0.0   
99                                    Woodbine Heights  0.0   
100                                    York Mills West  0.0   

     Accessories Store  Adult Boutique  American Restaurant  Antique Shop  \
0                  0.0             0.0                 0.03           0.0   
1                  0.0             0.0                 0.00           0.0   
2                  0.0             0.0                 0.00           0.0   
3                  0.0             0.0                 0.00           0.0   
4                  0.0             0.0                 0.00           0.0   
..                 ...             ...                  ...           ...   
96                 0.0             0.0                 0.00           0.0   
97                 0.0             0.0                 0.00           0.0   
98                 0.0             0.0                 0.00           0.0   
99                 0.0             0.0                 0.00           0.0   
100                0.0             0.0                 0.00           0.0   

     Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  ...  \
0         0.0         0.01         0.0                  0.0  ...   
1         0.0         0.00         0.0                  0.0  ...   
2         0.0         0.00         0.0                  0.0  ...   
3         0.0         0.00         0.0                  0.0  ...   
4         0.0         0.00         0.0                  0.0  ...   
..        ...          ...         ...                  ...  ...   
96        0.0         0.00         0.0                  0.0  ...   
97        0.0         0.00         0.0                  0.0  ...   
98        0.0         0.00         0.0                  0.0  ...   
99        0.0         0.00         0.0                  0.0  ...   
100       0.0         0.00         0.0                  0.0  ...   

     Vegetarian / Vegan Restaurant  Veterinarian  Video Game Store  \
0                             0.01           0.0               0.0   
1                             0.00           0.0               0.0   
2                             0.00           0.0               0.0   
3                             0.00           0.0               0.0   
4                             0.00           0.0               0.0   
..                             ...           ...               ...   
96                            0.00           0.0               0.0   
97                            0.00           0.0               0.0   
98                            0.00           0.0               0.0   
99                            0.00           0.0               0.0   
100                           0.00           0.0               0.0   

     Video Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  \
0         0.0000                    0.0      0.01        0.0          0.0   
1         0.0000                    0.0      0.00        0.0          0.0   
2         0.0000                    0.0      0.00        0.0          0.0   
3         0.0625                    0.0      0.00        0.0          0.0   
4         0.0000                    0.0      0.00        0.0          0.0   
..           ...                    ...       ...        ...          ...   
96        0.0000                    0.0      0.00        0.0          0.0   
97        0.0000                    0.0      0.00        0.0          0.0   
98  

<span style="color:darkred">*Finally we create a dataframe showing the 5 most common venues for each neighborhood.*</span>

In [17]:
# function that returns the 5 most common venues for each neighborhood (row)
def return_most_common_venues(row):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:5]

# create columns
df_columns = ['Neighborhood Name']
for i in np.arange(5):
    df_columns.append('#{} Most Common Venue'.format(i+1))

# create a new dataframe
venues_sorted = pd.DataFrame(columns=df_columns)
venues_sorted['Neighborhood Name'] = venues_avg['Neighborhood Name']
for i in np.arange(venues_avg.shape[0]):
    venues_sorted.iloc[i, 1:] = return_most_common_venues(venues_avg.iloc[i, :])

# show results
venues_sorted.head()

Neighborhood Name #1 Most Common Venue  \
0                           Adelaide, King, Richmond          Coffee Shop   
1                                          Agincourt   Chinese Restaurant   
2  Agincourt North, L'Amoreaux East, Milliken, St...             Pharmacy   
3  Albion Gardens, Beaumond Heights, Humbergate, ...        Grocery Store   
4                             Alderwood, Long Branch    Convenience Store   

   #2 Most Common Venue   #3 Most Common Venue #4 Most Common Venue  \
0                  Café                  Hotel           Restaurant   
1  Hong Kong Restaurant         Discount Store      Bubble Tea Shop   
2          Intersection            Yoga Studio                 Farm   
3              Pharmacy             Beer Store  Fried Chicken Joint   
4                   Gym  Performing Arts Venue                  Pub   

  #5 Most Common Venue  
0  Japanese Restaurant  
1         Skating Rink  
2    Elementary School  
3  Japanese Restaurant  
4    Elementary School

##### <span style="color:darkred">3- Run k-Means algorithm' nearby venues</span>

<span style="color:darkred">*Run k-Means algorithm with 4 clusters*</span>

In [18]:
from sklearn.cluster import KMeans

venues_clustering = venues_avg.drop('Neighborhood Name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=NB_CLUSTERS, random_state=0).fit(venues_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2, 3, 2, 2, 0, 2, 0, 2, 2,
       2, 2, 0, 2, 0, 2, 2, 2, 0, 3, 0, 0, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 0, 2, 2, 2, 0, 2, 0, 3, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2,
       2, 1, 0, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2])

<span style="color:darkred">*Add clustering and geospacial information to the sorted dataframe*</span>

In [19]:
# add clustering labels
venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = df.join(venues_sorted.set_index('Neighborhood Name'), on='Neighborhood')

# show results
toronto_merged

PostalCode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood  Latitude  Longitude  \
0                                       Rouge, Malvern  43.81139  -79.19662   
1               Highland Creek, Rouge Hill, Port Union  43.78574  -79.15875   
2                    Guildwood, Morningside, West Hill  43.76575  -79.17470   
3                                               Woburn  43.76812  -79.21761   
4                                            Cedarbrae  43.76944  -79.23892   
..                                                 ...       ...        ...   
98                                              Weston  43.70507  -79.51804   
99                                           Westmount  43.69630  -79.52926   
100  Kingsview Village, Martin Grove Gardens, Richv...  43.68681  -79.55728   
101  Albion Gardens, Beaumond Heights, Humbergate, ...  43.74453  -79.58624   
102                                          Northwest  43.71174  -79.57941   

     Cluster Labels        #1 Most Common Venue       #2 Most Common Venue  \
0               3.0        Fast Food Restaurant                Yoga Studio   
1               3.0  Construction & Landscaping                        Bar   
2               0.0  Construction & Landscaping                       Park   
3               2.0            Business Service      Korean BBQ Restaurant   
4               2.0                 Gaming Cafe                      Trail   
..              ...                         ...                        ...   
98              2.0                 Pizza Place                       Park   
99              2.0                 Pizza Place                Coffee Shop   
100             2.0                 Pizza Place                   Bus Line   
101             2.0               Grocery Store                   Pharmacy   
102             2.0                  Restaurant  Middle Eastern Restaurant   

          #3 Most Common Venue  #4 Most Common Venue  #5 Most Common Venue  
0            Electronics Store                  Food           Flower Shop  
1               Farmers Market           Escape Room  Ethiopian Restaurant  
2                     Bus Stop  Gym / Fitness Center     Fish & Chips Shop  
3                         Park           Coffee Shop                  Farm  
4                  Yoga Studio     Elementary School           Escape Room  
..                         ...                   ...                   ...  
98                       Diner   Fried Chicken Joint     Convenience Store  
99   Middle Eastern Restaurant        Sandwich Place    Chinese Restaurant  
100             Clothing Store           Music Venue              Bus Stop  
101                 Beer Store   Fried Chicken Joint   Japanese Restaurant  
102             Sandwich Place      Storage Facility                   Gym  

[103 rows x 11 columns]

<span style="color:darkred">*Analyze the number of neighborhoods per cluster*</span>

In [20]:
toronto_merged['Cluster Labels'].value_counts()

2.0    81
0.0    13
3.0     4
1.0     3
Name: Cluster Labels, dtype: int64

<span style="color:darkred">*Group the 2 neighborhoods with no nearby venues together in an additional cluster*</span>

In [21]:
print('There are {} unassigned neighborhoods.'.format(toronto_merged['Cluster Labels'].isnull().sum()))
toronto_merged['Cluster Labels'].fillna(NB_CLUSTERS, inplace=True)
print('There are {} unassigned neighborhoods.'.format(toronto_merged['Cluster Labels'].isnull().sum()))
toronto_merged['Cluster Labels'].value_counts()

There are 2 unassigned neighborhoods.
There are 0 unassigned neighborhoods.


2.0    81
0.0    13
3.0     4
1.0     3
4.0     2
Name: Cluster Labels, dtype: int64

<span style="color:darkred">*Visualize the clusters on the map*</span>

In [22]:
# recreate and center map
map_toronto = folium.Map(location=[loc.latitude, loc.longitude], zoom_start=11)

# define color range
colors_array = cm.rainbow(np.linspace(0, 1, NB_CLUSTERS+1))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# iterate on boroughs
for cluster in range(NB_CLUSTERS+1):
    add_markers_to_map(toronto_merged[toronto_merged['Cluster Labels']==cluster].reset_index(),
        rainbow[cluster], "Cluster='{}'".format(cluster), map_toronto)
    
# show map
map_toronto

##### <span style="color:darkred">4- Analyze clusters</span>

<span style="color:darkred">*Let's display the neighborhoods details cluster by cluster to spot the common types of venues*</span>

In [23]:
toronto_merged[toronto_merged['Cluster Labels']==0]

PostalCode           Borough  \
2         M1E       Scarborough   
17        M2H        North York   
20        M2L        North York   
25        M3A        North York   
34        M4A        North York   
40        M4J         East York   
46        M4R   Central Toronto   
50        M4W  Downtown Toronto   
64        M5P   Central Toronto   
79        M6L        North York   
82        M6P      West Toronto   
88        M8V         Etobicoke   
90        M8X         Etobicoke   

                                     Neighborhood  Latitude  Longitude  \
2               Guildwood, Morningside, West Hill  43.76575  -79.17470   
17                              Hillcrest Village  43.80225  -79.35558   
20                       Silver Hills, York Mills  43.75698  -79.38060   
25                                      Parkwoods  43.75245  -79.32991   
34                               Victoria Village  43.73057  -79.31306   
40                                   East Toronto  43.68811  -79.33418   
46                             North Toronto West  43.71458  -79.40668   
50                                       Rosedale  43.68190  -79.37829   
64            Forest Hill North, Forest Hill West  43.69479  -79.41440   
79             Downsview, North Park, Upwood Park  43.71381  -79.48874   
82                  High Park, The Junction South  43.65973  -79.46281   
88   Humber Bay Shores, Mimico South, New Toronto  43.60987  -79.49809   
90  The Kingsway, Montgomery Road, Old Mill North  43.65319  -79.51113   

    Cluster Labels        #1 Most Common Venue  \
2              0.0  Construction & Landscaping   
17             0.0                        Park   
20             0.0                        Park   
25             0.0           Food & Drink Shop   
34             0.0           German Restaurant   
40             0.0           Convenience Store   
46             0.0                  Playground   
50             0.0                        Park   
64             0.0                   Locksmith   
79             0.0                      Bakery   
82             0.0           Convenience Store   
88             0.0                        Park   
90             0.0                        Pool   

                        #2 Most Common Venue #3 Most Common Venue  \
2                                       Park             Bus Stop   
17  Residential Building (Apartment / Condo)          Yoga Studio   
20                               Yoga Studio       Farmers Market   
25                                      Park       Farmers Market   
34                                      Park        Grocery Store   
40                              Intersection                 Park   
46                                      Park             Gym Pool   
50                                Playground           Bike Trail   
64                                      Park          Yoga Studio   
79                          Basketball Court                 Park   
82                            Sandwich Place                 Park   
88                               Yoga Studio        Grocery Store   
90                                      Park          Yoga Studio   

                        #4 Most Common Venue #5 Most Common Venue  
2                       Gym / Fitness Center    Fish & Chips Shop  
17                                      Farm    Elementary School  
20                         Elementary School          Escape Room  
25                         Elementary School          Escape Room  
34                               Yoga Studio   Falafel Restaurant  
40                                      Farm    Elementary School  
46                               Yoga Studio                 Farm  
50                               Yoga Studio                 Farm  
64                            Farmers Market          Escape Room  
79                            Farmers Market          Escape Room  
82  Residential Building (Apartment / Condo)          Event

In [24]:
toronto_merged[toronto_merged['Cluster Labels']==1]

PostalCode     Borough                                     Neighborhood  \
28        M3H  North York  Bathurst Manor, Downsview North, Wilson Heights   
81        M6N        York                    The Junction North, Runnymede   
96        M9L  North York                                    Humber Summit   

    Latitude  Longitude  Cluster Labels #1 Most Common Venue  \
28  43.75788  -79.44847             1.0         Home Service   
81  43.67646  -79.48272             1.0              Brewery   
96  43.75948  -79.55707             1.0         Home Service   

   #2 Most Common Venue #3 Most Common Venue    #4 Most Common Venue  \
28     Business Service    Electronics Store             Flower Shop   
81         Home Service   Seafood Restaurant  Furniture / Home Store   
96           Hobby Shop  Rental Car Location      Falafel Restaurant   

   #5 Most Common Venue  
28          Flea Market  
81                 Farm  
96    Electronics Store

In [25]:
toronto_merged[toronto_merged['Cluster Labels']==2]

PostalCode      Borough  \
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood  Latitude  Longitude  \
3                                               Woburn  43.76812  -79.21761   
4                                            Cedarbrae  43.76944  -79.23892   
5                                  Scarborough Village  43.74446  -79.23117   
6          East Birchmount Park, Ionview, Kennedy Park  43.72582  -79.26461   
7                      Clairlea, Golden Mile, Oakridge  43.71289  -79.28506   
..                                                 ...       ...        ...   
98                                              Weston  43.70507  -79.51804   
99                                           Westmount  43.69630  -79.52926   
100  Kingsview Village, Martin Grove Gardens, Richv...  43.68681  -79.55728   
101  Albion Gardens, Beaumond Heights, Humbergate, ...  43.74453  -79.58624   
102                                          Northwest  43.71174  -79.57941   

     Cluster Labels #1 Most Common Venue       #2 Most Common Venue  \
3               2.0     Business Service      Korean BBQ Restaurant   
4               2.0          Gaming Cafe                      Trail   
5               2.0                  Spa          Indian Restaurant   
6               2.0    Convenience Store         Chinese Restaurant   
7               2.0         Intersection                     Bakery   
..              ...                  ...                        ...   
98              2.0          Pizza Place                       Park   
99              2.0          Pizza Place                Coffee Shop   
100             2.0          Pizza Place                   Bus Line   
101             2.0        Grocery Store                   Pharmacy   
102             2.0           Restaurant  Middle Eastern Restaurant   

          #3 Most Common Venue #4 Most Common Venue #5 Most Common Venue  
3                         Park          Coffee Shop                 Farm  
4                  Yoga Studio    Elementary School          Escape Room  
5                   Restaurant                 Park        Grocery Store  
6               Discount Store          Coffee Shop     Department Store  
7                  Coffee Shop         Soccer Field             Bus Line  
..                         ...                  ...                  ...  
98                       Diner  Fried Chicken Joint    Convenience Store  
99   Middle Eastern Restaurant       Sandwich Place   Chinese Restaurant  
100             Clothing Store          Music Venue             Bus Stop  
101                 Beer Store  Fried Chicken Joint  Japanese Restaurant  
102             Sandwich Place     Storage Facility                  Gym  

[81 rows x 11 columns]

In [26]:
toronto_merged[toronto_merged['Cluster Labels']==3]

PostalCode      Borough                                    Neighborhood  \
0         M1B  Scarborough                                  Rouge, Malvern   
1         M1C  Scarborough          Highland Creek, Rouge Hill, Port Union   
32        M3M   North York                               Downsview Central   
80        M6M         York  Del Ray, Keelesdale, Mount Dennis, Silverthorn   

    Latitude  Longitude  Cluster Labels        #1 Most Common Venue  \
0   43.81139  -79.19662             3.0        Fast Food Restaurant   
1   43.78574  -79.15875             3.0  Construction & Landscaping   
32  43.73224  -79.50178             3.0  Construction & Landscaping   
80  43.69517  -79.48397             3.0                         ATM   

    #2 Most Common Venue #3 Most Common Venue #4 Most Common Venue  \
0            Yoga Studio    Electronics Store                 Food   
1                    Bar       Farmers Market          Escape Room   
32      Insurance Office                 Farm    Elementary School   
80  Fast Food Restaurant           Playground          Coffee Shop   

          #5 Most Common Venue  
0                  Flower Shop  
1         Ethiopian Restaurant  
32                 Escape Room  
80  Construction & Landscaping

<span style="color:darkred">*Analyzing the common venues per cluster, we can highlight the following patterns:*</span><br>
<span style="color:darkred">- *Cluster __0__ regroups the 13 places having __parks__ in their neighborhood*</span><br>
<span style="color:darkred">- *Cluster __1__ regroups the 3 places having a high number of __home services__ in their neighborhood*</span> <br>
<span style="color:darkred">- *Cluster __3__ regroups the 4 places having __Construction / Landscaping venues__ in their neighborhood*</span> <br>
<span style="color:darkred">- *Cluster __2__ regroups all the other places, which have more __restaurants and other food venues__ in their direct neighborhood*</span> 